{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "78980e91-c7f3-40f0-a7f1-9b83447fb527",
   "metadata": {},
   "source": [
    "# Comparing alignment-free and usual MSA-based phylogenetic matrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70e05633-53d5-43a9-bf7a-f10982c10257",
   "metadata": {},
   "source": [
    "## Constructing our dataset of alpha, beta, delta and gammacoronaviruses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3607abc8-b71b-4260-a5c5-9cd2e8c21df3",
   "metadata": {},
   "source": [
    "### Downloading genomic sequences from NCBI via custon E-utilities custom link\n",
    "Sequences obtained following https://doi.org/10.1371/journal.pone.0264640.\n",
    "Use it in a simple web-browser."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be8b630e-f5b6-420a-930b-cf230f75571b",
   "metadata": {},
   "source": [
    "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id=NC_022103.1,NC_048216.1,NC_046964.1,NC_032107.1,NC_028833.1,NC_028824.1,NC_028814.1,NC_028811.1,NC_032730.1,NC_018871.1,NC_025217.1,NC_030886.1,MN996532,NC_034440.1,NC_014470.1,NC_010438.1,NC_010437.1,NC_009988.1,NC_009021.1,NC_009020.1,NC_009019.1,NC_009657.1,NC_005831.2,NC_002645.1,NC_045512.2,NC_006213.1,NC_038294.1,NC_019843.3,NC_006577.2,NC_004718.3,NC_034972.1,NC_028752.1,MT121216.1,MT040336.1,NC_039207.1,NC_026011.1,NC_017083.1,NC_012936.1,NC_003045.1,KX432213.1,JX860640.1,NC_010646.1,NC_038861.1,NC_030292.1,NC_028806.1,NC_023760.1,NC_002306.3,NC_003436.1,NC_039208.1,NC_010800.1,NC_048214.1,NC_048213.1,NC_046965.1,NC_001451.1,NC_011547.1,NC_016992.1,NC_016991.1,NC_016996.1,NC_016995.1,NC_016994.1,NC_016993.1,NC_011550.1,NC_011549.1,NC_001846.1,NC_048217.1,AC_000192.1&rettype=fasta\n",
    "\n",
    "\n",
    "Then, place the file sequences in the notebook's folder in order to facilitate finding this file.\n",
    "(We named this file as cov4gen.fasta)\n",
    "__________________________________________________________________________________________________________"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10465ef1-57da-4637-917f-ed658db056a2",
   "metadata": {},
   "source": [
    "Verifying its content:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "400e792a-23f3-4316-bf2f-b18d0a273ed2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "66\n"
     ]
    }
   ],
   "source": [
    "!grep -c \">\" cov4gen.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "815cb8a2-159b-46e8-a295-f54f2b0f398f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_022103.1 Bat coronavirus CDPHE15/USA/2006, complete genome\n",
      ">NC_048216.1 NL63-related bat coronavirus strain BtKYNL63-9b, complete genome\n",
      ">NC_046964.1 Alphacoronavirus Bat-CoV/P.kuhlii/Italy/3398-19/2015, complete genome\n",
      ">NC_032107.1 NL63-related bat coronavirus strain BtKYNL63-9a, complete genome\n",
      ">NC_028833.1 BtNv-AlphaCoV/SC2013, complete genome\n",
      ">NC_028824.1 BtRf-AlphaCoV/YN2012, complete genome\n",
      ">NC_028814.1 BtRf-AlphaCoV/HuB2013, complete genome\n",
      ">NC_028811.1 BtMr-AlphaCoV/SAX2011, complete genome\n",
      ">NC_032730.1 Lucheng Rn rat coronavirus isolate Lucheng-19, complete genome\n",
      ">NC_018871.1 Rousettus bat coronavirus HKU10, complete genome\n",
      ">NC_025217.1 Bat Hp-betacoronavirus/Zhejiang2013, complete genome\n",
      ">NC_030886.1 Rousettus bat coronavirus isolate GCCDC1 356, complete genome\n",
      ">MN996532.2 Bat coronavirus RaTG13, complete genome\n",
      ">NC_034440.1 Bat coronavirus isolate PREDICT/PDF-2180, complete genome\n",
      ">NC_014470.1 Bat coronavirus BM48-31/BGR/2008, complete genome\n",
      ">NC_010438.1 Miniopterus bat coronavirus HKU8, complete genome\n",
      ">NC_010437.1 Bat coronavirus 1A, complete genome\n",
      ">NC_009988.1 Rhinolophus bat coronavirus HKU2, complete genome\n",
      ">NC_009021.1 Rousettus bat coronavirus HKU9, complete genome\n",
      ">NC_009020.1 Pipistrellus bat coronavirus HKU5, complete genome\n",
      ">NC_009019.1 Tylonycteris bat coronavirus HKU4, complete genome\n",
      ">NC_009657.1 Scotophilus bat coronavirus 512, complete genome\n",
      ">NC_005831.2 Human Coronavirus NL63, complete genome\n",
      ">NC_002645.1 Human coronavirus 229E, complete genome\n",
      ">NC_045512.2 Severe acute respiratory syndrome coronavirus 2 isolate Wuhan-Hu-1, complete genome\n",
      ">NC_006213.1 Human coronavirus OC43 strain ATCC VR-759, complete genome\n",
      ">NC_038294.1 Betacoronavirus England 1 isolate H123990006, complete genome\n",
      ">NC_019843.3 Middle East respiratory syndrome-related coronavirus isolate HCoV-EMC/2012, complete genome\n",
      ">NC_006577.2 Human coronavirus HKU1, complete genome\n",
      ">NC_004718.3 SARS coronavirus Tor2, complete genome\n",
      ">NC_034972.1 Coronavirus AcCoV-JC34, complete genome\n",
      ">NC_028752.1 Camel alphacoronavirus isolate camel/Riyadh/Ry141/2015, complete genome\n",
      ">MT121216.1 Pangolin coronavirus isolate MP789, complete genome\n",
      ">MT040336.1 Pangolin coronavirus isolate PCoV_GX-P5E, complete genome\n",
      ">NC_039207.1 Betacoronavirus Erinaceus/VMC/DEU/2012 isolate ErinaceusCoV/2012-174/GER/2012, complete genome\n",
      ">NC_026011.1 Betacoronavirus HKU24 strain HKU24-R05005I, complete genome\n",
      ">NC_017083.1 Rabbit coronavirus HKU14, complete genome\n",
      ">NC_012936.1 Rat coronavirus Parker, complete genome\n",
      ">NC_003045.1 Bovine coronavirus isolate BCoV-ENT, complete genome\n",
      ">KX432213.1 Canine respiratory coronavirus strain BJ232, complete genome\n",
      ">JX860640.1 Canine respiratory coronavirus strain K37, complete genome\n",
      ">NC_010646.1 Beluga Whale coronavirus SW1, complete genome\n",
      ">NC_038861.1 Transmissible gastroenteritis virus complete genome, genomic RNA\n",
      ">NC_030292.1 Ferret coronavirus isolate FRCoV-NL-2010, complete genome\n",
      ">NC_028806.1 Swine enteric coronavirus strain Italy/213306/2009, complete genome\n",
      ">NC_023760.1 Mink coronavirus strain WD1127, complete genome\n",
      ">NC_002306.3 Feline infectious peritonitis virus, complete genome\n",
      ">NC_003436.1 Porcine epidemic diarrhea virus, complete genome\n",
      ">NC_039208.1 Porcine coronavirus HKU15 strain HKU15-155, complete genome\n",
      ">NC_010800.1 Turkey coronavirus, complete genome\n",
      ">NC_048214.1 Duck coronavirus isolate DK/GD/27/2014, complete genome\n",
      ">NC_048213.1 Infectious bronchitis virus isolate Ind-TN92-03, complete genome\n",
      ">NC_046965.1 Canada goose coronavirus strain Cambridge_Bay_2017, complete genome\n",
      ">NC_001451.1 Avian infectious bronchitis virus, complete genome\n",
      ">NC_011547.1 Bulbul coronavirus HKU11-934, complete genome\n",
      ">NC_016992.1 Sparrow coronavirus HKU17, complete genome\n",
      ">NC_016991.1 White-eye coronavirus HKU16, complete genome\n",
      ">NC_016996.1 Common moorhen coronavirus HKU21, complete genome\n",
      ">NC_016995.1 Wigeon coronavirus HKU20, complete genome\n",
      ">NC_016994.1 Night-heron coronavirus HKU19, complete genome\n",
      ">NC_016993.1 Magpie-robin coronavirus HKU18, complete genome\n",
      ">NC_011550.1 Munia coronavirus HKU13-3514, complete genome\n",
      ">NC_011549.1 Thrush coronavirus HKU12-600, complete genome\n",
      ">NC_001846.1 Murine hepatitis virus strain MHV-A59 C12 mutant, complete genome\n",
      ">NC_048217.1 Murine hepatitis virus strain A59, complete genome\n",
      ">AC_000192.1 Murine hepatitis virus strain JHM, complete genome\n"
     ]
    }
   ],
   "source": [
    "!grep \">\" cov4gen.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c41a281a-cab3-4ec6-9ff6-45d26d3679ff",
   "metadata": {},
   "source": [
    "We add these three sequences from GISAID, as described in the method from the above paper:\n",
    "* Beijing.IVDC_01\n",
    "* Beijing.IVDC_02\n",
    "* Beijing.IVDC_03"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c07adeba-aef9-410c-92de-ae7d91a979f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat cov4gen.fasta 1673264835172.sequences.fasta > db_cov.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cc317776-487b-4c05-9495-d71087744d43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "69\n"
     ]
    }
   ],
   "source": [
    "!grep -c \">\" db_cov.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "880525ca-03d0-46ca-8115-74c09a52b4da",
   "metadata": {},
   "source": [
    "### Multiple Sequence Analysis\n",
    "Alignment of the 69 sequences using mafft linsi method. \n",
    "(We do not recommend to run this in a common user PC because of its wide RAM usage).\n",
    "This command above consumed high ammounts of RAM!!! (+to 60 GB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3553ff1c-cdcb-489d-afaa-7ee962cbc5ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "!nohup linsi --thread 32 --leavegappyregion db_cov.fasta > alig_cov.fasta &"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fff3f169-2f64-4e5c-bb6b-5ca08270bfdb",
   "metadata": {},
   "source": [
    "Before running IQTREE, we trimmed the alignment with trimAl v1.4.rev22. We choose to remove all positions in the alignment with gaps in 10% or more of the sequences, unless this leaves less than 70% of original alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a55fd2e7-1b95-445d-a5ce-68a91ff8f7f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!trimal -in alig_cov.fasta -out trimm_alig_cov.fasta -gt 0.9 -cons 70"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7c9b09a-5162-499a-83eb-1a2fc9cee4f6",
   "metadata": {},
   "source": [
    "### Estimating Maximum-Likelihood Phylogeny\n",
    "The maximum-likelihood phylogenetic trees were obtained using IQTREE v2.2.0 software with the Ultrafast Bootstrap parameter set at 1000 and 1000 replicates of the SH-aLRT test. We used the GTR+I+G substitution model, as there is evidence that this model choice leads to phylogenetic analyses comparable to those made using current strategies for evolutionary model selection."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c26dfb89-e118-4a1a-8832-adee1d8ef5d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "IQTREE2=/dados/software/iqtree-2.2.0-Linux/bin/iqtree2\n",
    "$IQTREE2 -T AUTO -s trimm_alig_cov.fasta -m GTR+I+G -B 1000 -alrt 1000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7022ea2d-5264-4cde-a5e1-2cf9b0a05ae4",
   "metadata": {},
   "source": [
    "## Installing used R packages\n",
    "### One must properly install rpy2 before continue\n",
    "https://rpy2.github.io/doc/latest/html/overview.html#install-installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a9a155a3-5cd7-4fda-8a7a-e1f53851853e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b204cf6e-6a15-40dd-9075-ca6c1dde26d8",
   "metadata": {},
   "source": [
    "Instalation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d5c60021-1550-4ee1-89ae-56bc2e3e4d74",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: 'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "R[write to console]: Bioconductor version 3.15 (BiocManager 1.30.20), R 4.2.2 Patched (2022-11-10\n",
      "  r83330)\n",
      "\n",
      "R[write to console]: Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, foreign, Matrix, mgcv, spatial\n",
      "\n",
      "R[write to console]: 'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
      "Replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "R[write to console]: Bioconductor version 3.15 (BiocManager 1.30.20), R 4.2.2 Patched (2022-11-10\n",
      "  r83330)\n",
      "\n",
      "R[write to console]: Installation paths not writeable, unable to update packages\n",
      "  path: /usr/lib/R/library\n",
      "  packages:\n",
      "    boot, foreign, Matrix, mgcv, spatial\n",
      "\n",
      "R[write to console]: Installing packages into ‘/home/murilo/R/x86_64-pc-linux-gnu-library/4.2’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/tidyverse_2.0.0.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 704618 bytes (688 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 688 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/ape_5.7.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 1577932 bytes (1.5 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 1.5 MB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/tidytree_0.4.2.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 64216 bytes (62 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 62 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/ggpubr_0.6.0.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 1941659 bytes (1.9 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 1.9 MB\n",
      "\n",
      "\n",
      "* installing *source* package ‘tidyverse’ ...\n",
      "** package ‘tidyverse’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (tidyverse)\n",
      "* installing *source* package ‘ape’ ...\n",
      "** package ‘ape’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c BIONJ.c -o BIONJ.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c NNI.c -o NNI.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c RcppExports.cpp -o RcppExports.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c SPR.c -o SPR.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c additive.c -o additive.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c ape.c -o ape.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bNNI.c -o bNNI.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bionjs.c -o bionjs.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bipartition.c -o bipartition.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c bitsplits.c -o bitsplits.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c delta_plot.c -o delta_plot.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c dist_dna.c -o dist_dna.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c dist_nodes.c -o dist_nodes.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c ewLasso.c -o ewLasso.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c heap.c -o heap.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c mat_expo.c -o mat_expo.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c me.c -o me.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c me_balanced.c -o me_balanced.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c me_ols.c -o me_ols.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c mvr.c -o mvr.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c mvrs.c -o mvrs.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c nj.c -o nj.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c njs.c -o njs.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c pic.c -o pic.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c plot_phylo.c -o plot_phylo.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c prop_part.cpp -o prop_part.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c rTrait.c -o rTrait.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c read_dna.c -o read_dna.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c reorder_Rcpp.cpp -o reorder_Rcpp.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c reorder_phylo.c -o reorder_phylo.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c treePop.c -o treePop.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c tree_build.c -o tree_build.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c tree_phylo.c -o tree_phylo.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c triangMtd.c -o triangMtd.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c triangMtds.c -o triangMtds.o\n",
      "gcc -I\"/usr/share/R/include\" -DNDEBUG  -I'/home/murilo/R/x86_64-pc-linux-gnu-library/4.2/Rcpp/include'    -fpic  -g -O2 -ffile-prefix-map=/build/r-base-Faorqz/r-base-4.2.2.20221110=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -c ultrametric.c -o ultrametric.o\n",
      "g++ -std=gnu++14 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -flto=auto -ffat-lto-objects -flto=auto -Wl,-z,relro -o ape.so BIONJ.o NNI.o RcppExports.o SPR.o additive.o ape.o bNNI.o bionjs.o bipartition.o bitsplits.o delta_plot.o dist_dna.o dist_nodes.o ewLasso.o heap.o mat_expo.o me.o me_balanced.o me_ols.o mvr.o mvrs.o nj.o njs.o pic.o plot_phylo.o prop_part.o rTrait.o read_dna.o reorder_Rcpp.o reorder_phylo.o treePop.o tree_build.o tree_phylo.o triangMtd.o triangMtds.o ultrametric.o -llapack -lblas -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/murilo/R/x86_64-pc-linux-gnu-library/4.2/00LOCK-ape/00new/ape/libs\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (ape)\n",
      "* installing *source* package ‘ggpubr’ ...\n",
      "** package ‘ggpubr’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "*** moving datasets to lazyload DB\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (ggpubr)\n",
      "* installing *source* package ‘tidytree’ ...\n",
      "** package ‘tidytree’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (tidytree)\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: \n",
      "R[write to console]: The downloaded source packages are in\n",
      "\t‘/tmp/RtmpxOb2Y9/downloaded_packages’\n",
      "R[write to console]: \n",
      "R[write to console]: \n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "# Install and load the Bioconductor manager\n",
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "  install.packages(\"BiocManager\")\n",
    "BiocManager::install()\n",
    "\n",
    "# Install the required Bioconductor packages\n",
    "BiocManager::install(c(\"phyloTop\", \"treeio\", \"castor\", \"ggtree\"))\n",
    "\n",
    "# Install the required CRAN packages\n",
    "install.packages(c(\"tidyverse\", \"ape\", \"ggtree\", \"tidytree\", \"ggpubr\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b663bbae-d077-4b97-9da0-d79a6c33ae35",
   "metadata": {},
   "source": [
    "Load the installed packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "354a58ab-1d27-4b8b-9d69-87b0e992e5a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Error: package or namespace load failed for ‘tidyverse’:\n",
      " .onAttach failed in attachNamespace() for 'tidyverse', details:\n",
      "  call: value[[3L]](cond)\n",
      "  error: Package ‘ggplot2’ version 3.4.0 cannot be unloaded:\n",
      " Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘patchwork’, ‘aplot’, ‘ggplotify’, ‘ggtree’, ‘ggfun’, ‘ggpmisc’, ‘ggpp’ so cannot be unloaded\n",
      "\n",
      "R[write to console]: In addition: \n",
      "R[write to console]: Warning messages:\n",
      "\n",
      "R[write to console]: 1: package(s) not installed when version(s) same as or greater than current; use\n",
      "  `force = TRUE` to re-install: 'phyloTop' 'treeio' 'castor' 'ggtree' \n",
      "\n",
      "R[write to console]: 2: package ‘ggtree’ is not available for this version of R\n",
      "\n",
      "A version of this package for your version of R might be available elsewhere,\n",
      "see the ideas at\n",
      "https://cran.r-project.org/doc/manuals/r-patched/R-admin.html#Installing-packages \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Error: package or namespace load failed for ‘tidyverse’:\n",
      " .onAttach failed in attachNamespace() for 'tidyverse', details:\n",
      "  call: value[[3L]](cond)\n",
      "  error: Package ‘ggplot2’ version 3.4.0 cannot be unloaded:\n",
      " Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘patchwork’, ‘aplot’, ‘ggplotify’, ‘ggtree’, ‘ggfun’, ‘ggpmisc’, ‘ggpp’ so cannot be unloaded\n"
     ]
    },
    {
     "ename": "RInterpreterError",
     "evalue": "Failed to parse and evaluate line 'library(phyloTop)\\nlibrary(treeio)\\nlibrary(castor)\\nlibrary(tidyverse)\\nlibrary(ape)\\nlibrary(ggtree)\\n'.\nR error message: \"Error: package or namespace load failed for ‘tidyverse’:\\n .onAttach failed in attachNamespace() for 'tidyverse', details:\\n  call: value[[3L]](cond)\\n  error: Package ‘ggplot2’ version 3.4.0 cannot be unloaded:\\n Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘patchwork’, ‘aplot’, ‘ggplotify’, ‘ggtree’, ‘ggfun’, ‘ggpmisc’, ‘ggpp’ so cannot be unloaded\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRRuntimeError\u001b[0m                             Traceback (most recent call last)",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:331\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    329\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    330\u001b[0m     \u001b[38;5;66;03m# Need the newline in case the last line in code is a comment.\u001b[39;00m\n\u001b[0;32m--> 331\u001b[0m     value, visible \u001b[38;5;241m=\u001b[39m \u001b[43mro\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mr\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mwithVisible(\u001b[39;49m\u001b[38;5;124;43m{\u001b[39;49m\u001b[38;5;132;43;01m%s\u001b[39;49;00m\u001b[38;5;130;43;01m\\n\u001b[39;49;00m\u001b[38;5;124;43m})\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m%\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    332\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    333\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/robjects/__init__.py:459\u001b[0m, in \u001b[0;36mR.__call__\u001b[0;34m(self, string)\u001b[0m\n\u001b[1;32m    458\u001b[0m p \u001b[38;5;241m=\u001b[39m rinterface\u001b[38;5;241m.\u001b[39mparse(string)\n\u001b[0;32m--> 459\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mp\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    460\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m conversion\u001b[38;5;241m.\u001b[39mget_conversion()\u001b[38;5;241m.\u001b[39mrpy2py(res)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/robjects/functions.py:204\u001b[0m, in \u001b[0;36mSignatureTranslatedFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    203\u001b[0m         kwargs[r_k] \u001b[38;5;241m=\u001b[39m v\n\u001b[0;32m--> 204\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m (\u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mSignatureTranslatedFunction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    205\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/robjects/functions.py:127\u001b[0m, in \u001b[0;36mFunction.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    126\u001b[0m         new_kwargs[k] \u001b[38;5;241m=\u001b[39m cv\u001b[38;5;241m.\u001b[39mpy2rpy(v)\n\u001b[0;32m--> 127\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mFunction\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[38;5;21;43m__call__\u001b[39;49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_args\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mnew_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    128\u001b[0m res \u001b[38;5;241m=\u001b[39m cv\u001b[38;5;241m.\u001b[39mrpy2py(res)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/rinterface_lib/conversion.py:45\u001b[0m, in \u001b[0;36m_cdata_res_to_rinterface.<locals>._\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_\u001b[39m(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs):\n\u001b[0;32m---> 45\u001b[0m     cdata \u001b[38;5;241m=\u001b[39m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     46\u001b[0m     \u001b[38;5;66;03m# TODO: test cdata is of the expected CType\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/rinterface.py:817\u001b[0m, in \u001b[0;36mSexpClosure.__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    816\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m error_occured[\u001b[38;5;241m0\u001b[39m]:\n\u001b[0;32m--> 817\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m embedded\u001b[38;5;241m.\u001b[39mRRuntimeError(_rinterface\u001b[38;5;241m.\u001b[39m_geterrmessage())\n\u001b[1;32m    818\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m res\n",
      "\u001b[0;31mRRuntimeError\u001b[0m: Error: package or namespace load failed for ‘tidyverse’:\n .onAttach failed in attachNamespace() for 'tidyverse', details:\n  call: value[[3L]](cond)\n  error: Package ‘ggplot2’ version 3.4.0 cannot be unloaded:\n Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘patchwork’, ‘aplot’, ‘ggplotify’, ‘ggtree’, ‘ggfun’, ‘ggpmisc’, ‘ggpp’ so cannot be unloaded\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mRInterpreterError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_cell_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mR\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlibrary(phyloTop)\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mlibrary(treeio)\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mlibrary(castor)\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mlibrary(tidyverse)\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mlibrary(ape)\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mlibrary(ggtree)\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/IPython/core/interactiveshell.py:2362\u001b[0m, in \u001b[0;36mInteractiveShell.run_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2360\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m   2361\u001b[0m     args \u001b[38;5;241m=\u001b[39m (magic_arg_s, cell)\n\u001b[0;32m-> 2362\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2363\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:870\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    868\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m e\u001b[38;5;241m.\u001b[39mstdout\u001b[38;5;241m.\u001b[39mendswith(e\u001b[38;5;241m.\u001b[39merr):\n\u001b[1;32m    869\u001b[0m         \u001b[38;5;28mprint\u001b[39m(e\u001b[38;5;241m.\u001b[39merr)\n\u001b[0;32m--> 870\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    871\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    872\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice \u001b[38;5;129;01min\u001b[39;00m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpng\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msvg\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:850\u001b[0m, in \u001b[0;36mRMagics.R\u001b[0;34m(self, line, cell, local_ns)\u001b[0m\n\u001b[1;32m    848\u001b[0m         return_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    849\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 850\u001b[0m     text_result, result, visible \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43meval\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcode\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    851\u001b[0m     text_output \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m text_result\n\u001b[1;32m    852\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m visible:\n",
      "File \u001b[0;32m~/.local/lib/python3.10/site-packages/rpy2/ipython/rmagic.py:335\u001b[0m, in \u001b[0;36mRMagics.eval\u001b[0;34m(self, code)\u001b[0m\n\u001b[1;32m    332\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (ri\u001b[38;5;241m.\u001b[39membedded\u001b[38;5;241m.\u001b[39mRRuntimeError, \u001b[38;5;167;01mValueError\u001b[39;00m) \u001b[38;5;28;01mas\u001b[39;00m exception:\n\u001b[1;32m    333\u001b[0m     \u001b[38;5;66;03m# Otherwise next return seems to have copy of error.\u001b[39;00m\n\u001b[1;32m    334\u001b[0m     warning_or_other_msg \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflush()\n\u001b[0;32m--> 335\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m RInterpreterError(code, \u001b[38;5;28mstr\u001b[39m(exception),\n\u001b[1;32m    336\u001b[0m                             warning_or_other_msg)\n\u001b[1;32m    337\u001b[0m text_output \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mflush()\n\u001b[1;32m    338\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m text_output, value, visible[\u001b[38;5;241m0\u001b[39m]\n",
      "\u001b[0;31mRInterpreterError\u001b[0m: Failed to parse and evaluate line 'library(phyloTop)\\nlibrary(treeio)\\nlibrary(castor)\\nlibrary(tidyverse)\\nlibrary(ape)\\nlibrary(ggtree)\\n'.\nR error message: \"Error: package or namespace load failed for ‘tidyverse’:\\n .onAttach failed in attachNamespace() for 'tidyverse', details:\\n  call: value[[3L]](cond)\\n  error: Package ‘ggplot2’ version 3.4.0 cannot be unloaded:\\n Error in unloadNamespace(package) : namespace ‘ggplot2’ is imported by ‘patchwork’, ‘aplot’, ‘ggplotify’, ‘ggtree’, ‘ggfun’, ‘ggpmisc’, ‘ggpp’ so cannot be unloaded\""
     ]
    }
   ],
   "source": [
    "%%R\n",
    "library(phyloTop)\n",
    "library(treeio)\n",
    "library(castor)\n",
    "library(tidyverse)\n",
    "library(ape)\n",
    "library(ggtree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4d0bd05-0a1c-4129-a021-d7448d986a0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d99912d-a5ce-4d98-b3eb-2ea2374a5b81",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ffd1217-b420-4864-9a04-0ef19a27850c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "047ef88f-3600-4247-abd7-154e4f120c3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd36be75-102c-4767-b769-506de5af55d3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d165bec0-1e46-4602-a41f-855c69f5c53e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa311931-3648-40d1-81b7-ffc554d76fa9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
